independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/RestClientRequestContext.java

```diff
@@ -202,7 +203,14 @@ public HttpClientResponse getVertxClientResponse() {
     @Override
     protected Executor getEventLoop() {
         if (httpClientRequest == null) {
-            return restClient.getVertx().nettyEventLoopGroup().next();
+            // make sure we execute the client callbacks on the same context as the current thread
```

Comment:
This is not right, if this is being called from a worker then it will dispatch on the worker. You still need to actually dispatch to the relevant IO thread.

Reply:
How can this be called from a worker thread, though? I've always seen it called from an IO thread.

Reply:
It could be called by the user in org.jboss.resteasy.reactive.client.spi.ResteasyReactiveClientRequestContext from a org.jboss.resteasy.reactive.client.spi.ResteasyReactiveClientRequestFilter, which could be called from pretty much anywhere. 

TBH I am really worried about this 'swapping threads' approach. It opens the door to lots of subtle bugs as we now need to ensure that we always do some kind of safe hand off between the threads. We also need to be super careful with IO, as things work differently if you are not on your IO thread, because as soon as you resume IO its possible another IO thread has started running the input handler, so you need to thing about the thread safety semantics of that.

From a performance point of view it also sucks, as it will involve multiple IO threads waking up selectors that don't belong to them, which is terrible from a performance POV.

Reply:
I agree with the concerns, there's many similar difficulties with Hibernate Reactive.

As a user, I think I'd be fine to have to deal with stronger restrictions in exchange for less [potential] trouble.

Reply:
Maybe we should just merge this for now (once the tests are passing), but look to the possibility of moving to a single thread model in future.

Thinking about it I am not sure how practical it will be, as having basically a pool of connections per IO thread is not going to be great for databases which have connection limits.

Reply:
I think we should merge it for now, as I said above. I have rebased so lets see what CI says.